### PR TITLE
PE-6794: IO Usage Events

### PR DESCRIPTION
**Pull request by arielmelendez**
Added tracking events for:
- Transfer
- Create-Vault
- Delegate-Stake
- Decrease-Delegate-Stake
- Cancel-Delegate-Withdrawl
- "Extend-Vault"
- "Increase-Vault"

Added function `eventingPcall` that DRYs out sending errors along with events when pcalls fail.

### CODE REVIEW

**Review by atticusofsparta — COMMENTED** [review]
Reccomend adding integration tests for events using AoLoader

File: src/main.lua [hunk 1]
```diff
@@ -193,40 +230,53 @@ Handlers.add(ActionMap.CreateVault, utils.hasMatchingTag("Action", ActionMap.Cre
...
+		ioEvent:addField("VaultEndTimestamp", vault.endTimestamp)
 	end
+
+	ao.send({
+		Target = msg.From,
+		Tags = {
+			Action = "Vault-Created-Notice",
+			["Vault-Id"] = msg.Id,
+		},
+		Data = json.encode(shouldContinue2),
```
Comment (arielmelendez) [inline src/main.lua:277]: This should prooooobably be `vault` (json encoded).

File: src/main.lua [hunk 2]
```diff
@@ -292,34 +344,49 @@ Handlers.add(ActionMap.ExtendVault, utils.hasMatchingTag("Action", ActionMap.Ext
...
 			Target = msg.From,
 			Tags = { Action = "Invalid-Extend-Vault-Notice", Error = "Invalid-Extend-Vault" },
-			Data = tostring(err),
-		})
-	else
-		ao.send({
-			Target = msg.From,
-			Tags = { Action = "Vault-Extended-Notice" },
-			Data = json.encode(result),
+			Data = tostring(error),
```
Comment (arielmelendez) [inline src/main.lua:364]: `error` should probably be json encoded `vault`.

File: src/main.lua [hunk 3]
```diff
@@ -693,87 +775,119 @@ Handlers.add(ActionMap.DelegateStake, utils.hasMatchingTag("Action", ActionMap.D
...
-			ao.send({
-				Target = msg.From,
-				Tags = {
-					Action = "Invalid-Cancel-Delegate-Withdrawl-Notice",
-					Error = "Invalid-Cancel-Delegate-Withdrawl",
-				},
-				Data = tostring(result),
-			})
-		else
+		local shouldContinue2, result = eventingPcall(ioEvent, function(error)
```
Comment (arielmelendez) [inline src/main.lua:846]: Rename `result` to `gateway`.

File: src/main.lua (hunk 1 above)
Comment (arielmelendez) [reply src/main.lua]: 9e21524

File: src/ao_event.lua [hunk 4]
```diff
@@ -0,0 +1,115 @@
+-- Factory function for creating an "AOEvent"
+local function AOEvent(initialData)
```
Comment (dtfiedler) [inline src/ao_event.lua:2]: i haven't seen lua classes created this way - the most common pattern i've seen is more like

```
-- Define a class
local MyClass = {}
MyClass.__index = MyClass

-- Constructor for the class
function MyClass:new(name, age)
    local instance = setmetatable({}, MyClass) -- Set the metatable for the new object
    instance.name = name
    instance.age = age
    return instance
end

-- A method for the class
function MyClass:greet()
    print("Hello, my name is " .. self.name .. " and I am " .. self.age .. " years old.")
end

-- Create an instance of the class
local person = MyClass:new("John", 30)
person:greet()  -- Output: Hello, my name is John and I am 30 years old.
```

but regardless - we can look to use classes more for certain objects (gateways, records) to put some protections around accessing/modifying properties directly)

File: src/gar.lua [hunk 5]
```diff
@@ -705,6 +706,10 @@ function gar.cancelDelegateWithdrawal(from, gatewayAddress, vaultId)
 	delegate.delegatedStake = delegate.delegatedStake + vaultBalance
 	gateway.totalDelegatedStake = gateway.totalDelegatedStake + vaultBalance
 	GatewayRegistry[gatewayAddress] = gateway
+	return {
+		delegate = gar.getGateway(gatewayAddress).delegates[from],
+		totalDelegatedStake = gateway.totalDelegatedStake
+	}
```
Comment (dtfiedler) [inline src/gar.lua:712]: inconsequential change - but required a modification to the unit test that validates the result is no linger `nil`

**Review by dtfiedler — APPROVED** [review]
can get this into devnet and test against a local cu/honeycomb - if there's an issue, our CI/CD and monitoring tests https://github.com/ar-io/ar-io-network-process/actions/workflows/monitor.yaml

should catch broken handlers/unexpected state

**Pull request merged**